ocaml/xapi/rfb_client_msgtype_parser.ml

```diff
@@ -0,0 +1,268 @@
+  let parse_client_cut_text =
+    take 1 >>= fun msg_type ->
+    if Char.code msg_type.[0] = 6 then
+      take 3 >>= fun _ ->
+      Angstrom.BE.any_int32 >>= fun text_length ->
+      take (Int32.to_int text_length) >>| fun _ -> ClientCutText
+    else
+      fail __FUNCTION__
+
+  (* Parse QEMU Client Message: message-type(1) + submessage-type(1) + data(10) = 12 bytes *)
+  let parse_qemu_client_message =
+    take 1 >>= fun msg_type ->
+    if Char.code msg_type.[0] = 255 then
+      take 11 >>| fun _ -> QEMUClientMessage
+    else
+      fail __FUNCTION__
+
+  (* Fallback parser for unknown messages *)
+  let parse_unsupported_message =
+    take 1 >>= fun msg_type ->
+    let hex_data = hex_dump_data msg_type 1 in
+    (* Use commit to prevent backtracking, so we can fail and get the error message *)
+    commit *> fail ("UnsupportedMsg: " ^ hex_data)
+
+  let handshake_parser = parse_handshake
+
+  let message_parser =
+    (* Put the most likely parsers first *)
+    choice
+      [
+        parse_framebuffer_update_request
+      ; parse_pointer_event
+      ; parse_key_event
+      ; parse_client_cut_text
+      ; parse_qemu_client_message
+      ; parse_set_pixel_format
+      ; parse_set_encodings
+      ; parse_unsupported_message (* Fallback *)
+      ]
+
+  (* Create RFB parser with closure-based state encapsulation *)
+  (* An example usage of the parser:
+     let rfb_parser = create () in
+     fun data ->
+       try
+         let parsed_messages = rfb_parser data in
+         Handle_messages
+       with exn ->
+         Error_handler
+  *)
+  let create () =
+    let module State = struct
+      type t = {
+          parser_state: msg Angstrom.Buffered.state
+        ; parser_type: msg Angstrom.t
+      }
+    end in
+    (* Private state hidden in closure *)
+    let state =
+      ref
+        {
+          State.parser_state= Angstrom.Buffered.parse handshake_parser
+        ; State.parser_type= handshake_parser
+        }
+    in
+
+    (* Helper to get unconsumed data as string *)
+    let unconsumed_to_string {Angstrom.Buffered.buf; off; len} =
+      Bigstringaf.substring buf ~off ~len
+    in
+
+    (* Decide next parser type (handshake_parser or message_parser) *)
+    let update_parser_type parsed_result =
+      let new_parser =
+        match parsed_result with
+        | Handshake ->
+            message_parser
+        | _ ->
+            !state.State.parser_type
+      in
+      state :=
+        {
+          State.parser_state= !state.State.parser_state
+        ; State.parser_type= new_parser
+        }
+    in
+
+    let rec check_parsing_result acc =
+      let open Angstrom.Buffered in
+      Result.bind acc @@ fun (parser_state, msgs) ->
+      match parser_state with
+      | Done (unconsumed, parsed_result) ->
+          let msgs = parsed_result :: msgs in
+          (* Update parser type based on parsed result *)
+          update_parser_type parsed_result ;
```

Comment:
The `State.parser_type` is not necessary. The whole handshake must have completed when it reaches here. So within this function, only the `message_parser` will be used.